WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/handlers/StartupUIThread.java

```diff
@@ -20,29 +20,30 @@
  */
 public class StartupUIThread implements Runnable {
 
-	private final StartUpHandler startUpHandler;
-
 	/** The preferences. */
 	private Preferences preferences;
 
+	/** Whether user has cancelled first registration. */
+	private boolean userRegistrationCancelled = false;
+
 	/** Constructor. */
-	public StartupUIThread(StartUpHandler startUpHandler,
-			Preferences preferences) {
-		this.startUpHandler = startUpHandler;
+	public StartupUIThread(Preferences preferences) {
 		this.preferences = preferences;
 	}
 
 	@Override
 	public void run() {
 		// the execution strategy is:
-		// (1) check user registration (2) wait until 1 is complete, or 1
-		// can be skipped (3) show workspace registration
+		// (1) check user registration (2) wait until user+project registration
+		// is completed, or 1 can be skipped (3) if user is already registered,
+		// check for workspace registration
 		if (WatchDogUtils.isEmpty(preferences.getUserid())) {
 			displayUserRegistrationWizard();
 		}
 
-		if (!WatchDogUtils.isEmpty(preferences.getUserid())) {
-			// In case the user aborted the preference dialog with cancel,
+		if (!WatchDogUtils.isEmpty(preferences.getUserid())
```

Comment:
userRegistrationCancelled is a boolean flag that I've added. It ensures that if the user cancelled wizard after registering user-id, but before registering project, he would not be asked again for project registration (if condition false, ignoring the block).Otherwise, if he changes workspace, he will be offered just to do workspace registration (if condition true, going into block). Is my explanation clear enough?